### ISSUE TITLE
PHP cannot find SMTP class even though SMTP.php is included

### ISSUE THREAD

**Issue by David263** [PHPMailer/PHPMailer#2360]
Please check these things before submitting your issue:

- [ ] *Read the error message* you're seeing - it often tells you what is wrong, and may contain useful links & instructions
- [ ] Make sure you're using the latest version of PHPMailer
- [ ] Check that your problem is not dealt with in [the troubleshooting guide](https://github.com/PHPMailer/PHPMailer/wiki/Troubleshooting), especially if you're having problems connecting to Gmail or GoDaddy
- [ ] Include sufficient code to reproduce your problem
- [ ] If you're having an SMTP issue, include the debug output generated with `SMTPDebug = 2` set
- [ ] If you have a question about how to use PHPMailer (rather than reporting a bug in it), tag a question on Stack Overflow with `phpmailer`, but [**search first**](http://stackoverflow.com/questions/tagged/phpmailer)!

# Problem description

I have tried several times in the past to understand PHP Namespaces and have failed, so I accept that I will never understand how they work. This may be a simple Namespace problem.

In trying to get the simple PHPMailer > $mail->send() example working, I'm stuck on the first error message: `Fatal error: Uncaught Error: Class 'SMTP' not found`. I highly doubt that this could really be a bug in a mature product like PHPMailer, but I need help at this point. Note that my code includes the SMTP.php file, yet the SMTP class is undefined. Strange.

# Code to reproduce

`// I put this code inside a function:
use PHPMailer\PHPMailer\PHPMailer;
use PHPMailer\PHPMailer\Exception;

require 'PHPMailer/src/Exception.php';
require 'PHPMailer/src/PHPMailer.php';
require 'PHPMailer/src/SMTP.php'; // Yes, this exists

// . . .

$mail = new PHPMailer(true);

try {
		//Server settings
		$mail->SMTPDebug = SMTP::DEBUG_SERVER;
		$mail->isSMTP(); 
		$mail->Host       =$MailServer; 
		$mail->SMTPAuth   = true; 
		$mail->Username   = $SendingAcct; 
		$mail->Password   = $SendingPwd;
		$mail->SMTPSecure = PHPMailer::ENCRYPTION_SMTPS; 
		$mail->Port       = 465; 

		//Recipients
		$mail->setFrom($FromAddr, $FromName);
		$mail->addAddress($ToAddr, $ToName);

		//Attachments

		//Content
		$mail->isHTML(true);
		$mail->Subject=$Subject;
		$mail->Body=$Body;
		$mail->AltBody=$Body;

		$mail->send();
		echo 'Message has been sent';
		} catch (Exception $e) {
		echo "Message could not be sent. Mailer Error: {$mail->ErrorInfo}";
		}
`

# Debug output

**Comment by Synchro** [comment]
Easy mistake! It is a simple namespace problem.

The problem is this line

    $mail->SMTPDebug = SMTP::DEBUG_SERVER;

In order to get at that constant, you need to import the SMTP class into your (global) namespace, so add:

    use PHPMailer\PHPMailer\SMTP;

at the top of the script.

The reason you need that is that the SMTP class is namespaced, so you need to `use` it to import that class name into your namespace. You could achieve the same by using a fully-qualified class name (FQCN) like this:

    $mail->SMTPDebug = \PHPMailer\PHPMailer\SMTP::DEBUG_SERVER;

Occasionally that format is useful, but it's usually neater to go with `use` statements. It took me a while to get used to PHP's namespaces; you'll get there in the end!

**Comment by David263** [comment]
Thank you! Do you think it is really worth using strange Namespaces and 'use' for a simple library like PHPMailer? If so, can anyone point me to a good tutorial? I absolutely cannot understand what's in the PHP Manual.

For my own libraries, I use a short name for the library and class, if any, like J for JSON. Then the function names are like this: J_FromJSON, J_ToJSON. Very simple and keeps the library names unique. In JavaScript I use a different mechanism (nested functions), so names look like this: J.FromJSON, J.ToJSON (I use lowercase for variable names, capitalize function names).

**Comment by Synchro** [comment]
"Naming things" is a classically difficult thing in comp sci!
While you can stay on top of names in small projects, in bigger ones it becomes much more difficult. Having namespaces as part of the language makes it much easier to be consistent and organised about it, and there is no conflict between approaches because everyone uses the same system. Much of composer and packagist (and thus modern PHP development in general) would not be practical without namespaces.

Fundamentally PHP namespaces are quite simple. All a `use` statement does is say, "in this script, call this class by this name".

As for their utility in small projects, namespaces and composer start paying off very quickly. In this very simple example you posted, because you're not using composer, you're already having to load 3 files manually instead of just 1, and when it comes to updating your libraries, it's going to take you much longer to do manually. Now imagine trying to manage an app with 1,000 dependencies!

**Comment by David263** [comment]
If I had to maintain many code dependencies I would use technology appropriate for the task, not the horribly unintuitive Namespaces of PHP. But, in fact, most of what I do is create websites. For this I develop my own minimal frameworks. Not everyone has the same goals! Designing for the biggest possible projects is overkill. Frameworks and technologies should adapt to the needs of the developer, not force obscure semantics on them.

But I will take your word for it that Namespaces are a good thing. Since you could not point to an easy-to-understand tutorial, I see a disconnect between advocacy of Namespaces and an unwillingness to explain them clearly.

I have tried to learn Namespaces from the Manual and from a tutorial, and still have only the vague idea that it is like a file system with directories and files. For my own projects, I don't want the overhead of all these "use" statements for every class I might want. I am content to use whatever names are used in a library and hide my own names were necessary. JavaScript works much better, in that it is easy to put names inside a function and thus hide them from anywhere outside. Too bad you can't have nested functions in PHP. Oh, wait a minute, you can indeed have nested functions in PHP!

**Comment by Synchro** [comment]
I didn't try looking for tutorials. That's not what I'm here for, and do not characterise this as me being "unwilling", which is frankly insulting. I'm sure you know how to use a search engine.

No you can't have nested functions in PHP. You might think you can (and it even looks like you can), but they are always global.

Namespaces are not necessarily anything to do with folders, though PSR-4 defines a naming convention to map from one to the other. Once you understand them, PHP namespaces are trivially simple to work with, and they are very simple and intuitive to work with. They are nothing more than local aliases, so I don't see why you're so deeply confused by them. The 300,000 packages on packagist are a straightforward testament to that.

**Comment by David263** [comment]
I've had my first success with PHPMailer. I suggest documenting that the hostname and probably the username too must be lowercase when dealing with a mail server hosted on Linux. This isn't obvious to a Windows programmer like me. Also, PHPMailer fails when using an aliased domain for the hostname, which I very much want to do. Maybe I can figure out a workaround in cPanel or the DNS zone record.

**Comment by Synchro** [comment]
The only thing that PHPMailer cares about when it comes to the `Host` property is whether it resolves in your DNS, which is well outside its scope. PHPMailer can't tell if a domain is an alias or anything else, it just asks the PHP runtime to resolve it.

**Comment by David263** [comment]
I apologize for insulting you. As I said, I have found Namespaces very hard to understand. No, it is not your job to explain them to me, even though your product uses them and gave me a problem because of them. I hope I do find a tutorial that can make them seem simple, but it is not top priority for me with a website to complete.

**Comment by David263** [comment]
I solved my aliased host problem, which wasn't an aliased host problem. It was a problem with the sending account. Changing the sending account ("Username") to the real host name worked fine even though the hostname, from, and to fields all used the aliased name (as desired).

**Comment by David263** [comment]
I was surprised to see `charset=iso-8859-1` in the sent email from your first example. https://stackoverflow.com/questions/2491475/phpmailer-character-encoding-issues states that one should add `$mail->CharSet = 'UTF-8';` . You might want to change your example code wherever it appears, as UTF-8 is now the encoding standard and default pretty much everywhere.

**Comment by Synchro** [comment]
I would love to do that, but handling UTF-8 requires the mbstring PHP extension, which is not guaranteed to be available (it's not part of PHP's base installation), so it can't be the default.

**Comment by David263** [comment]
I've written quite a bit of UTF-8 text handling code in PHP over the past year or two and have found that mbstring is rarely needed. The reality is that PHP strings have always been able to contain arbitrary strings of bytes. PHP never assumes any meaning to those bytes in its string handling functions (that is, it assumes that a character is a byte). So you can pass UTF-8 strings freely inside of PHP. They will never be mangled.

Now, if you wanted to scan the characters (either staying in UTF-8 or as Unicode code points), then you can no longer view a string as a sequence of bytes. Then you have to use the mb_substr function.

As far as I can imagine, PHPMailer only has to look for one-byte characters, such as period, CR, and newline. These are guaranteed unique in any UTF-8 string, meaning that the the UTF-8 encoding for characters after the first in a character sequence can **never** be identical to a one-byte character. That is the amazing way that UTF-8 was designed: control characters are unchanged. (The other amazing features of UTF-8 are its ability to encode variable-length strings in any language, and its remarkable ability to re-synchronize immediately when a byte is lost in transmission).

In summary, I would not expect PHPMailer to require any special changes other than making UTF-8 the default name of the encoding method in the generated email headers. Test it out and see.

**Comment by Synchro** [comment]
It must be nice to be so blissfully unaware of the many constraints imposed by email. Strangely enough I also know how UTF-8 works. Email is not the web. In many situations email is not even 8-bit clean, let alone multi-byte capable. Headers for example do not support 8-bit charsets at all (unless you're using RFC6531, which is tricky and relatively new). You can represent UTF-8 via numerous encoding transformations such as RFC2047, and at the same time you have to take into account line length limits in both RFC 821 and 822. As a result, PHPMailer does indeed have to go very deep into inspecting every string, and taking into account multi-byte characters, ensuring they are not broken arbitrarily. On top of all that, if you have so much as a single bit out of place, DKIM signatures will fail. You can get by without mbstring if you're using 8-bit charsets such as ISO 8859, which is why it's PHPMailer's default. Search for uses of `mb*` functions in PHPMailer's code base; there are many, but they are only used if you enable UTF-8. It might be nice for *you* to switch your charset (and guess what, you can do just that when you create an instance!), but remember that there are probably 100 million web sites using it (it's the default mailer in WordPress) and many are not are exactly up to date, which is why PHPMailer needs to retain its default.

**Comment by David263** [comment]
Oh.
